library/Freckle/App/Stats.hs

```diff
@@ -0,0 +1,271 @@
+--       <$> ...
+--       <*> pure client
+--
+--     'runApp' app $ ...
+--   @
+--
+-- - Throughout your application code, emit metrics as desired
+--
+--   @
+--   import qualified Freckle.App.Stats as Stats
+--
+--   myFunction :: (MonadIO m, MonadReader env m, 'HasStatsClient' env) => m ()
+--   myFunction = do
+--     start <- liftIO getCurrentTime
+--     result <- myAction
+--
+--     Stats.'increment' \"action.attempt\"
+--     Stats.'histogramSinceMs' \"action.duration\" start
+--
+--     case result of
+--       Left err -> do
+--         Stats.'increment' \"action.failure\"
+--         -- ...
+--       Right x -. do
+--         Stats.'increment' \"action.success\"
+--         -- ...
+--   @
+--
+module Freckle.App.Stats
+  ( StatsSettings
+  , envParseStatsSettings
+
+  -- * Client
+  , StatsClient
+  , withStatsClient
+  , HasStatsClient(..)
+
+  -- * Reporting
+  , tagged
+  , increment
+  , counter
+  , gauge
+  , histogram
+  , histogramSince
+  , histogramSinceMs
+  ) where
+
+import Freckle.App.Prelude
+
+import Blammo.Logging
+import Control.Lens (Lens', lens, view, (&), (.~), (<>~))
+import Control.Monad.Reader (local)
+import Data.Aeson (Value(..))
+import Data.String
+import Data.Time (diffUTCTime)
+import Freckle.App.Ecs
+import qualified Freckle.App.Env as Env
+import qualified Network.StatsD.Datadog as Datadog
+
+data StatsSettings = StatsSettings
+  { amsEnabled :: Bool
+  , amsSettings :: Datadog.DogStatsSettings
+  , amsTags :: [(Text, Text)]
+  }
+
+envParseStatsSettings :: Env.Parser Env.Error StatsSettings
+envParseStatsSettings =
+  StatsSettings
+    <$> Env.switch "DOGSTATSD_ENABLED" mempty
+    <*> (buildSettings
+        <$> optional (Env.var Env.str "DOGSTATSD_HOST" mempty)
+        <*> optional (Env.var Env.auto "DOGSTATSD_PORT" mempty)
+        )
+    <*> (buildTags
+        <$> optional (Env.var Env.nonempty "DD_ENV" mempty)
+        <*> optional (Env.var Env.nonempty "DD_SERVICE" mempty)
+        <*> optional (Env.var Env.nonempty "DD_VERSION" mempty)
+        <*> Env.var Env.keyValues "DOGSTATSD_TAGS" (Env.def [])
+        )
+ where
+  buildSettings mHost mPort =
+    Datadog.defaultSettings
+      & maybe id (Datadog.host .~) mHost
+      . maybe id (Datadog.port .~) mPort
+
+  buildTags mEnv mService mVersion tags =
+    catMaybes
+        [ ("env", ) <$> mEnv
+        , ("environment", ) <$> mEnv -- Legacy
+        , ("service", ) <$> mService
+        , ("version", ) <$> mVersion
+        ]
+      <> tags
+
+data StatsClient = StatsClient
+  { amcClient :: Datadog.StatsClient
+  , amcTags :: [(Text, Text)]
+  }
+
+tagsL :: Lens' StatsClient [(Text, Text)]
+tagsL = lens amcTags $ \x y -> x { amcTags = y }
+
+class HasStatsClient env where
+  statsClientL :: Lens' env StatsClient
+
+instance HasStatsClient StatsClient where
+  statsClientL = id
+
+withStatsClient
+  :: (MonadMask m, MonadUnliftIO m)
+  => StatsSettings
+  -> (StatsClient -> m a)
+  -> m a
+withStatsClient StatsSettings {..} f = do
```

Comment:
This usage has a lot of benefits, but unfortunately we expect a `loadApp :: IO App` in a lot of places (e.g. `Test.withApp`), and you can't make such a function if you only have access to a `StatsClient` in a block like this.

I'll have to :thinking: on this.